### ISSUE TITLE
Add `Vector` data type to compiler and standard library

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#5030]
Add a new `Vector` data type which is fundamentally like `Array` but is immutable and hence does not have a region.

We should finish #4760 before starting on this ticket.

Each bullet below should probably be its own PR.

### Phase 1: Basics

- [x] Add `Vector.flix` with:
```
type alias Vector[t] = Array[t, true]
```

### Phase 2: Library

Add a few functions to the `Vector` module:

- [x] Add `Vector.nth`
- [x] Add `Vector.append`
- [x] Add `Vector.length`
- [x] Add `Vector.repeat`
- [x] Add `Vector.slice`
- [x] Add `Vector.map`

Internally these can use `Array` functions or have their own implementation, but they must have unit tests. 
(Can be copy-pasted from `TestArray`).

### Phase 3: Syntax

- [x] Add the `Vector` type to the Flix compiler itself and drop the enum.
- [ ] Add pattern matching in `Vector`'s. (We can reuse old code from `Array` which is still in the compiler).

### Phase 4: Use Vector and Expand Library

- [ ] Use `Vector` in the Datalog engine.
- [ ] Use `Vector` in the channel implementation.
- [x] Expand the `Vector` module with functions from the `Array` module.

**Comment by JonathanStarup** [comment]
I think maybe `Array[t, Static]` is better because then you can clearly see the casts instead of this magic pure array

**Comment by magnus-madsen** [comment]
> I think maybe `Array[t, Static]` is better because then you can clearly see the casts instead of this magic pure array. The beginning of this implementation is in Fixpoint/Tuple

But no writes would ever be supported, so there would be no effect.

**Comment by JonathanStarup** [comment]
I was just thinking about what to put into the vector `case Vector(...)`

**Comment by magnus-madsen** [comment]
Its a good point; I think really what we want is to disallow `Array[t, true]` (and similar) in `Safety`?

**Comment by JonathanStarup** [comment]
the counter point is that you need a cast to get this thing but yeah. it feels like the language assumes that its a region so true (and even false) is a bit weird

**Comment by magnus-madsen** [comment]
> the counter point is that you need a cast to get this thing but yeah. it feels like the language assumes that its a region so true (and even false) is a bit weird

Ultimately this will be changed. Its more of an artifact of how we do things today.

**Comment by JonathanStarup** [comment]
> `type alias Vector[t] = Array[t, true]`

Then anyone can just call `Array.put(2, 0, someVector)` Why not `opaque enum`?

**Comment by stephentetley** [comment]
Thanks @magnus-madsen - yes I'm happy to look into it.

**Comment by magnus-madsen** [comment]
> > `type alias Vector[t] = Array[t, true]`
> 
> Then anyone can just call `Array.put(2, 0, someVector)` (which will be pure). Why not `opaque enum`?

This is temporary. The idea is to add the `Vector` type to Flix itself.

**Comment by magnus-madsen** [comment]
And the reason not to use an opaque enum is because we want to use Flix to enforce the immutability.

**Comment by magnus-madsen** [comment]
@stephentetley Excellent.

I imagine the first version might require some casts, but then later, once we have compiler support those (hopefully few) casts can be removed.

**Comment by stephentetley** [comment]
I'm just over half way through - Array has a big API.

Happily I've managed with very few casts.

**Comment by JonathanStarup** [comment]
> And the reason not to use an opaque enum is because we want to use Flix to enforce the immutability.

I do see the point in doing this as the easiest first version, but it seems to me that opaque enum forces users to treat Vector as a Vector and nothing more, while the type alias allows users to observe the actual array, making the eventual proper type a breaking change.

**Comment by magnus-madsen** [comment]
If we use an opaque type then you have the wrap and unwrap it all the time. The point of the type alias is that it is trivial to remove it once `Vector` becomes a built-in type.

**Comment by stephentetley** [comment]
I've made a PR covering phase 1 and 2 plus the it adds rest of the pure array functions (part of phase 4).

#5420

**Comment by magnus-madsen** [comment]
Tasks left (as a note to myself):

- Add compiler support.
- Add instances for `Vector`.
- Rework Datalog engine to return a `Vector` of tuples (instead of list).
- Any cleanup (?)

**Comment by magnus-madsen** [comment]
Ill see if Jakob or myself can do the compiler-stuff.
